### PR TITLE
DEPR/API: disallow lists within list for set_index

### LINKED ISSUE #24046: API: capabilities of df.set_index
This is coming out of a discussion that has stalled #22225 (which is about adding `.set_index` to Series, see #21684). The discussion has shifted away from what capabilities a putative `Series.set_index` should have, but what capabilities `df.set_index` has *currently*.

The main issue (for @jreback) is that `df.set_index` takes arrays:
> > @jreback: There were several attempts to have DataFrame.set_index take an array as well, but these never got off the ground.

> @h-vetinari: I'm not sure *when*, but they certainly did get off the ground:

```
>>> import pandas as pd
>>> import numpy as np
>>> pd.__version__
'0.23.4'
>>>
>>> df = pd.DataFrame(np.random.randint(0, 10, (4, 4)), columns=list('abcd'))
>>> df.set_index(['a',          # label
...               df.index,     # Index
...               df.b ** 2,    # Series
...               df.b.values,  # ndarray
...               list('ABC'), # list
...               'c'])         # label again
              b  d
a   b      c
0 0 0  2 A 1  0  2
8 1 1  4 B 4  1  4
3 2 25 5 C 8  5  5
0 3 9  7 D 2  3  7
```

Further on:
> @jreback: @h-vetinari you are confusing the purpose of `.set_axis`. [...] The problem with `.set_index` on a DataFrame with an array is that it technically can work with an array and not keys. (meaning its not unambiguous)

I don't think I am confusing them. If I want to set the `.index`-attribute of a Series/DataFrame, then using `.set_index` is the most reasonable name by far. If anything, `set_axis` should be a superset of `set_index` (and a putative `set_columns`), that just switches between the two based on the `axis`-kwarg.

More than that, the current capabilities of `df.set_index` are a proper superset of `df.set_axis(axis=0)`**, in that it's possible to fill `keys` with *only* `Series`/`Index`/`ndarray`/`list` etc.:
```
>>> df.set_index(pd.Index(df.a))  # same result as Series directly below
>>> df.set_index(df.a) 
   a  b  c  d
a
0  0  0  1  2
8  8  1  4  4
3  3  5  8  5
0  0  3  2  7
>>> df.set_index(df.a.values)  # same result as list directly below
>>> df.set_index([[0, 8, 3, 0]])
   a  b  c  d
0  0  0  1  2
8  8  1  4  4
3  3  5  8  5
0  0  3  2  7
```
** there is one caveat, in that lists (and only lists; out of all containers) need to be wrapped in another list, i.e. `df.set_index([[0, 8, 3, 0]])` instead of `df.set_index([0, 8, 3, 0])`. This is the heart of the ambiguity that @jreback mentioned above.

Summing up:
* [x] `set_index` is the most natural name for setting the `.index`-attribute
* [x] `df.set_index` *should* be able to process list-likes (as it currently does; this is the source of the ambiguity of the list case).
* [ ] `df.set_axis` should be able to do everything that `df.set_index` does, and just switch between operating on index/columns based on the `axis`-kwarg (after all, `index` and `columns` are the two axes of a DF).
  * [ ] it could be considered to add a method `set_columns` on a `DataFrame` 
  * [ ] The `axis`-kwarg of `set_axis` should just switch between the behaviour of `set_index` (i.e. dealing with keys *and* array-likes) and `set_columns`.
* [ ] `Series.set_index` should support the same signature as `df.set_index`, with the exception of the `drop`-keyword (which only makes sense for column labels).
* [ ] For Series, the `set_index` and `set_axis` methods should be exactly the same.

Since I can't tag @pandas-dev/pandas-core, here are a few individual tags: @jreback @TomAugspurger @jorisvandenbossche @gfyoung @WillAyd @jbrockmendel @jschendel @toobaz.

### PR DESCRIPTION
**Pull request by h-vetinari**
- [x] one way to close #24046
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
- [x] add deprecation warning & test it

I wanted to add this before the RC gets cut for two reasons.
On the one hand, the docs have silently deprecated arrays in `df.set_index` compare [master](https://pandas-docs.github.io/pandas-docs-travis/api/generated/pandas.DataFrame.set_index.html):
> **keys** : *label or list of label*
> >   Name or names of the columns that will be used as the index.

and [0.23.4](https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.DataFrame.set_index.html):
> **keys** : *column label or list of column labels / arrays*

This is IMO a breaking change. I'd wait for the outcome of the discussion of #24046, but I feel that could easily fall under the table before the RC, so I wanted to provide a worked-out implementation.

Equally importantly,  #22486 added capabilities for lots of list-likes *within* a list to `df.set_index`, and has not seen a release yet. Therefore, deprecation would be much easier now, than after `0.24.0rc`.

@jreback @TomAugspurger @jorisvandenbossche @toobaz @WillAyd

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
>  On the one hand, the docs have silently deprecated arrays in df.set_index compare master:

I don't think we silently deprecated anything. There was a DOC PR (https://github.com/pandas-dev/pandas/pull/22775) updating the docstring, and I think it was rather an oversight of a contributor that didn't know that behaviour and which was not catched during review.

And that is a good catch! So I think the documentation change in this PR is certainly welcome, and the code changes require some discussion. Would you want to do the docstring changes in a separate PR that can be merged more quickly?

**Comment by jorisvandenbossche** [comment]
@h-vetinari and to give you some expectations: I think most of us are rather swamped with trying to get the rc out, so I am not sure we will have time to finalize the set_index related discussions before the RC

**Comment by h-vetinari** [comment]
@jorisvandenbossche 
> I don't think we silently deprecated anything. There was a DOC PR (#22775) updating the docstring, and I think it was rather an oversight of a contributor that didn't know that behaviour and which was not catched during review.

There was not accusation, I used "silent deprecation" to mean that this functionality is suddenly not mentioned anymore.

> @h-vetinari and to give you some expectations: I think most of us are rather swamped with trying to get the rc out, so I am not sure we will have time to finalize the set_index related discussions before the RC

Which is fair enough. That's why I put up this PR (even though I don't really have the time), to let you judge the trade-off (e.g. currently adding *more* list-likes due to #22486, vs. the ambiguity we're talking about in #24046).

**Comment by toobaz** [comment]
While I understand @jorisvandenbossche 's concern...

>  On the one hand, the docs have silently deprecated arrays in `df.set_index` compare [master](https://pandas-docs.github.io/pandas-docs-travis/api/generated/pandas.DataFrame.set_index.html):

... I think on aligning the docs to the functionalities we can all agree, and

>  #22486 added capabilities for lots of list-likes _within_ a list to `df.set_index`, and has not seen a release yet

I think this is an important argument even leaving aside #24046 , because I see (sorry for not noticing before) that tuples are among these "list-likes", **which I think is just wrong**. We discussed in several other issues the fact that in pandas we want to distinguish tuples - which are either keys of a ``MultiIndex``, or pairs of keys for the two axes of a ``DataFrame`` - from lists - which are collections.

So I think this PR really deserves to get considered for the release. Will review in detail later today.

**Comment by toobaz** [comment]
> it was an explicit review request to use `is_list_like` and not instance-checks for specific classes

I see... in a similar case I resolved to

https://github.com/pandas-dev/pandas/commit/32ee9732b823448b87848f6bcaefdc762868999c#diff-1e79abbbdd150d4771b91ea60a4e1cc7R2701

... but I agree this is annoying, so see #24702 .

**Comment by h-vetinari** [comment]
I saw @TomAugspurger mentioned cutting the RC soon. In the unlikely case that someone still wants to consider this on short notice, feel free to commit into this PR - I'm offline the next 6-7h.

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -1251,8 +1251,8 @@ Other API Changes
 - :class:`pandas.io.formats.style.Styler` supports a ``number-format`` property when using :meth:`~pandas.io.formats.style.Styler.to_excel` (:issue:`22015`)
 - :meth:`DataFrame.corr` and :meth:`Series.corr` now raise a ``ValueError`` along with a helpful error message instead of a ``KeyError`` when supplied with an invalid method (:issue:`22298`)
 - :meth:`shift` will now always return a copy, instead of the previous behaviour of returning self when shifting by 0 (:issue:`22397`)
-- :meth:`DataFrame.set_index` now allows all one-dimensional list-likes, raises a ``TypeError`` for incorrect types,
-  has an improved ``KeyError`` message, and will not fail on duplicate column names with ``drop=True``. (:issue:`22484`)
+- :meth:`DataFrame.set_index` now gives a better (and less frequent) KeyError, raises a ``ValueError`` for incorrect types,
+  and will not fail on duplicate column names with ``drop=True``. (:issue:`22484`)
```
Comment (h-vetinari) [inline doc/source/whatsnew/v0.24.0.rst]: After reverting the added list-likes from #22486, what's left are the three points from #22484 that the former was meant to address.

**Comment by jreback** [comment]
@h-vetinari would not be averse to: 1) improving the doc-string / examples 2) raising better errors on existing behavior, but the impl here does too many things.

Basically the non-controversial things would be fine.

**Comment by h-vetinari** [comment]
> @jreback: would not be averse to: 1) improving the doc-string / examples [...]

Split off #24762.

> [...] 2) raising better errors on existing behavior, but the impl here does too many things.

The chances here are as advertised, and *really* minimal. The only things worth mentioning are:
```diff
-         if not isinstance(keys, list):
-             keys = [keys]
+         if (is_scalar(keys) or isinstance(keys, tuple)
+                 or isinstance(keys, (ABCIndexClass, ABCSeries, np.ndarray))):
+             # make sure we have a container of keys/arrays we can iterate over
+             # tuples can appear as valid column keys!
+             keys = [keys]
+         elif not isinstance(keys, list):
+             raise ValueError(err_msg)
```

which replaces indiscriminate list-wrapping with a very reasonable list of allowed types, and

```diff
+         depr_warn = False
          for col in keys:
              if (is_scalar(col) or isinstance(col, tuple)) and col in self:
                  # if col is a valid column key, everything is fine
                  continue
              elif is_scalar(col) and col not in self:
                  # tuples that are not keys are not considered missing,
                  # but illegal (see below)
                  missing.append(col)
-             elif (not is_list_like(col, allow_sets=False)
-                   or getattr(col, 'ndim', 1) > 1):
-                 raise TypeError(...)
+             elif isinstance(col, list):
+                 depr_warn = True
+             elif (not isinstance(col, (ABCIndexClass, ABCSeries, np.ndarray))
+                   or getattr(col, 'ndim', 1) > 1):
+                 raise ValueError(err_msg)
```
which does the actual deprecation and restricts to the allowed types (the `TypeError` we added for #22486 is I think inconsistent with how this is usually handled via `ValueError` eleshwere).

The rest is just doc changes, removing some test cases added by #22486, and catching the deprecation warnings in the tests (without changing the tests themselves).

**Comment by TomAugspurger** [comment]
Sorry, I haven't been able to keep up. Are either of the following being deprecated?

```python
In [8]: df.set_index([['a', 'b', 'c']])
Out[8]:
   A  B
a  1  4
b  2  5
c  3  6

In [9]: df.set_index(['A', ['a', 'b', 'c']])
Out[9]:
     B
A
1 a  4
2 b  5
3 c  6
```

what's our recommendation to users doing this (the message should tell them).

**Comment by h-vetinari** [comment]
@TomAugspurger 
Both those cases are indeed being deprecated in this PR. As mentioned above lists are hardly ever used *as an array* (outside of toy example), and in most real-world use cases, the required data is already residing in Series/Index/ndarrays anyway. Furthermore, deprecating lists inside lists removed the ambiguity of `list_of_scalar` (keys vs values), and sacrifices no essential functionality as it's easy to wrap a list in an array.

I've adapted the warning message to give a clearer example of that.

**Comment by h-vetinari** [comment]
@TomAugspurger @jorisvandenbossche @toobaz @jreback 
friendly ping

**Comment by jreback** [comment]
not for 0.24.0

**Comment by h-vetinari** [comment]
@jreback @TomAugspurger @toobaz 
Not in a rush with this one in particular, but the diff after @24762 is now much smaller/simpler.

**Comment by h-vetinari** [comment]
@TomAugspurger @jorisvandenbossche @toobaz @jreback
Coming back to this after #25085 and #24984 are in.

**Comment by jorisvandenbossche** [comment]
@h-vetinari To make sure there is a common understanding: this PR deprecates the usage of *lists* for the case of *array-like of values* (so not for the *list* of labels/array-likes). 
So meaning: `df.set_index(['A', 'B'])` (list of labels) or `df.set_index([np.array(['a', 'b', 'c']])` (list of array-likes) are still allowed, but `df.set_index([['a', 'b', 'c']])` (list of array-likes where the array like is a list) would be deprecated?

Is there any ambiguous case this is solving? Or is it only with the intent to make it less complex/confusing for users?

**Comment by h-vetinari** [comment]
> @jorisvandenbossche: @h-vetinari To make sure there is a common understanding: this PR deprecates the usage of _lists_ for the case of _array-like of values_ (so not for the _list_ of labels/array-likes).

That's correct.

> Is there any ambiguous case this is solving? Or is it only with the intent to make it less complex/confusing for users?

One of the main points of discussion about the capabilities of `df.set_index` (e.g. in #24046) is that @jreback @WillAyd are strongly opposed to the fundamental ambiguity of `df.set_index(['a', 'b', 'c'])` (see e.g. [here](https://github.com/pandas-dev/pandas/issues/24046#issuecomment-451991224) or OP of #24046), because lists are overloaded as an array-like *and* a container for combining several index columns.

This PR tries to break the stalemate between longstanding capabilities (including mixing column labels and arrays since before v0.12.) and this ambiguity, by deprecating the array-like role of lists, which can functionally be easily replaced by Series/Index/np.ndarray.

This also improves the situation with the confusingly-similar-yet-totally-different
`df.set_index(['a', 'b', 'c'])` vs. `df.set_index([['a', 'b', 'c']])`.

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -79,6 +79,7 @@ Deprecations
 ~~~~~~~~~~~~
 
 - Deprecated the `M (months)` and `Y (year)` `units` parameter of :func: `pandas.to_timedelta`, :func: `pandas.Timedelta` and :func: `pandas.TimedeltaIndex` (:issue:`16344`)
+- :meth:`DataFrame.set_index` has deprecated using lists of values *within* lists. It remains possible to pass array-likes, both directly and within a list.
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: what would help here is a sub-section note that shows a good example of this.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by h-vetinari** [comment]
> @jreback: what would help here is a sub-section note that shows a good example of this.

I'm assuming you'll find the section I added too long, but it's easier to shorten than to add.

**Comment by h-vetinari** [comment]
@jreback 
Now green after the doc changes.

**Comment by h-vetinari** [comment]
@jreback 
PTAL at the doc section you requested

@jorisvandenbossche @TomAugspurger @gfyoung 
Feel free to chime in as well. :)

**Comment by jreback** [comment]
after some discussion in real space. This is introducing more of a special case. This is not desired as adds user complexitly w/o a corresponding improvement. Thanks for the try here. This is a pretty tricky area of api and don't want to make it more confusing.

**Comment by h-vetinari** [comment]
@jreback @jorisvandenbossche @TomAugspurger 
This is unfortunate - would have liked to make the case for it at the dev-chat. This point was also the principal (only?) complaint about introducing `Series.set_index` (#22225) with the same capabilities. How do we proceed there? (reminder: you closed that PR unilaterally, but the need of such a method for Series was not in question.)

PS. Lists might seem like a special case here (although #22264 did exactly the same for `.str.cat`), but it's IMO very important to start deprecating lists-as-arrays (as opposed to lists-as-collections) throughout the code base. That ambiguity is at the heart of a lot of issues.

**Pull request closed without merging**